### PR TITLE
Add cfgfile version 0.2.8.2.1

### PR DESCRIPTION
**Pull request by igormironchik**
Specify library name and version:  **cfgfile/0.2.8.2.1**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

I'm the author of this library.

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Croydon** [comment]
Why did you close/open so many times the same pull request? You know that you can push to a pull request new commits after having opened it, right?

**Comment by igormironchik** [comment]
First PR I created when I was not a member of Early Access Program, just wasn't sure that your Jenkins will start build checks. This is the only reason - I wasn't sure that CI will start build process. People are so people. Is it a problem for Conan Center Index that I created/closed so many PRs? I'm sorry for it.

**Comment by igormironchik** [comment]
> Add more commits by pushing to the master branch on igormironchik/conan-center-index.

Sure, I saw this and I knew that I can modify the same PR.

**Comment by Croydon** [comment]
No, it is fine now. Just so that you know in the future 🙂 

CI starts after people are added to the EAP too

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:47]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
```

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:4]:
```suggestion
```

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:53]: Should be called from package method

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:52]: Is this is header only what are you building?

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:62]: There's no export CMake targets or namespace for your project

```suggestion
```

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:57]: Seems odd to include https://github.com/igormironchik/cfgfile/blob/master/cfgfile/cfgfile.pri

please use `os.path.join`

```suggestion
        self.copy("*.hpp", src=os.path.join(self._source_subfolder, "cfgfile"), dst=os.path.join("include", "cfgfile"))
```

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:60]: Seems very unusual 🤔 Why was this required?

File: recipes/cfgfile/0.2.8.2.1/test_package/example.cpp
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/test_package/example.cpp:12]: Since the layout of a test_package is standard could we cut the complexity down?

Could we either
1. pass the path as a command line argument
2. just hardcode the `../test.cfg` since we always run in from `bin/`?

**Review by prince-chrismc — COMMENTED** [review]
Here's some suggestions and a few comments!

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:52]: Yes, this is header-only library, but for the convenience this library provides auxiliary generator of C++ code from declarative description of configuration file. This generator is auxiliary build tool.

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:47]: Nice suggestion, but 

```
tools.get(**self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
```

doesn't strip the root folder. I receive `source/source_subfolder/cfgfile-0.2.8.2.1-full/...` whereas `cfgfile-0.2.8.2.1-full` should be stripped.

Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:47]: Hmm, looks like something strange in my archive...

Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py]: If create archive with leading `./` your `strip_root=True` will not strip the root. I.e. with

```
tar -czvf cfgfile-0.2.8.2.1-full.tar.gz ./cfgfile-0.2.8.2.1-full
```

Conan will fail with strip. And with

```
tar -czvf cfgfile-0.2.8.2.1-full.tar.gz cfgfile-0.2.8.2.1-full
```

All is OK. Looks like a minor issue, doesn't it?

Comment (prince-chrismc) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py]: Why not just use the auto generated ones from GitHub? If you pass `./` then you created and extra subdirectory which explains why it failed

https://github.com/igormironchik/cfgfile/archive/refs/tags/0.2.8.2.1.tar.gz

File: recipes/cfgfile/0.2.8.2.1/conandata.yml
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conandata.yml]: You really should not change releases after publishing, changing the structure might break someone's scripts =)

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:51]: This should be in the build method, you can check the docs for good CMake examples https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md#how-to-provide-a-good-recipe

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:12]: Since you want to build

```suggestion
```

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:51]: Okay header only + tools?

you'll need our CMake wrapper, and you'll need to add back from settings https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-installer-packages-remove-some-settings-from-their-package-id

**Review by prince-chrismc — COMMENTED** [review]
So now I understand this is not header only (but headers and a tool) we need to make some adjustments!

Along with the clues you might want to check out this as well https://github.com/conan-io/conan-center-index/blob/master/docs/reviewing.md#cmake

File: recipes/cfgfile/0.2.8.2.1/conandata.yml
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conandata.yml]: I recreated archive to satisfy your `tools.get()` method. Nothing changed, just removed leading `./` that break your `strip_root=True`.

Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conandata.yml]: And `cfgfile version 0.2.8.2.1` is a special release, that was created specially for Conan, as PR will be merged no more changes in downloads will be made.

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:39]:
```suggestion
        self._cmake.configure(source_folder=os.path,join(self._source_subfolder + "generator"), build_folder=self._build_subfolder)
```

**Review by prince-chrismc — APPROVED** [review]

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (uilianries) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:14]:
```suggestion
    settings = "os", "arch"
```
Installer doesn't require compiler and build_type

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (uilianries) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:39]: It requires CMake wrapper.

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (uilianries) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:66]:
```suggestion
        cmake.install()

    def package_info(self):
        bin_path = os.path.join(self.package_folder, "bin")
        self.output.info("Appending PATH env var with : {}".format(bin_path))
        self.env_info.PATH.append(bin_path)
```

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (uilianries) [inline recipes/cfgfile/0.2.8.2.1/conanfile.py:38]:
```suggestion
        self._cmake = CMake(self)
        self._cmake.definitions["BUILD_EXAMPLES"] = False
        self._cmake.definitions["BUILD_TESTS"] = False
```

Do not build examples and test

**Review by uilianries — CHANGES_REQUESTED** [review]
Thanks for your contribution, take a look on my review please.

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:14]: Corresponding to Conan's documentation ([this one](https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-installer-packages-remove-some-settings-from-their-package-id)) `build_type` is required, as this library provides executable tool. Not sure about `compiler`, but use `compiler` setting in `vaildate` method. And I guided to `CMake` recipe...

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:39]: Could you specify what do you mean under CMake wrapper? I guess that I already use it.

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:66]: Possibly, but not sure that this is required. I successfully use `cfgfile.generator` in `test_package`... Why do we need to expand PATH environment here?

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:38]: I don't build tests and examples. I use `generator/CMakeLists.txt` that doesn't build any examples and tests. With this item I totally disagree.

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:39]: https://github.com/conan-io/conan-center-index/blob/master/recipes/acado/all/CMakeLists.txt

Is an example

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (prince-chrismc) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:66]: It's a convention we have to make it easier

Comment (uilianries) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:66]: test package uses run_environment ...

File: recipes/cfgfile/0.2.8.2.1/conanfile.py
Comment (uilianries) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:38]: Why don't you use the main cmake instead of walking to another cmake file? It would better using main and adding generator as optional, so `cmake --build . && cmake --build . --target install` is enough for all this

Comment (igormironchik) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:38]: Main CMake builds generator and use it as sub-folder, there I have options to build tests/examples. But for Conan recipe it's enough to use only generator's cmake. But if you want I can use main CMake and disable building of tests/examples with options.

Comment (uilianries) [reply recipes/cfgfile/0.2.8.2.1/conanfile.py:39]: True, but remove the compiler from package id: 

```
def package_id(self):
    del self.info.settings.compiler
```

It doesn't affect validate() method.

File: recipes/cfgfile/0.2.8.2.1/test_package/conanfile.py
Comment (uilianries) [inline recipes/cfgfile/0.2.8.2.1/test_package/conanfile.py:14]:
```suggestion
        if not tools.cross_building(self.settings):
            bin_path = os.path.join("bin", "cfgfile.test")
            self.run(bin_path, run_environment=True)
```

File: recipes/cfgfile/0.2.8.2.1/test_package/conanfile.py
Comment (uilianries) [inline recipes/cfgfile/0.2.8.2.1/test_package/conanfile.py:1]:
```suggestion
from conans import ConanFile, CMake, tools
```

**Review by uilianries — APPROVED** [review]

File: recipes/cfgfile/0.2.8.2.1/CMakeLists.txt
Comment (mathbunnyru) [inline recipes/cfgfile/0.2.8.2.1/CMakeLists.txt:2]:
```suggestion
cmake_minimum_required(VERSION 3.10)
```

File: recipes/cfgfile/0.2.8.2.1/test_package/CMakeLists.txt
Comment (mathbunnyru) [inline recipes/cfgfile/0.2.8.2.1/test_package/CMakeLists.txt:2]:
```suggestion
cmake_minimum_required(VERSION 3.1)
```

File: recipes/cfgfile/0.2.8.2.1/test_package/example.cpp
Comment (mathbunnyru) [inline recipes/cfgfile/0.2.8.2.1/test_package/example.cpp:2]:
```suggestion
#include <string>
```

File: recipes/cfgfile/0.2.8.2.1/test_package/example.cpp
Comment (mathbunnyru) [inline recipes/cfgfile/0.2.8.2.1/test_package/example.cpp:5]:
```suggestion
#include <cfg.hpp>
```

**Review by mathbunnyru — CHANGES_REQUESTED** [review]

**Review by mathbunnyru — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by jgsogo — APPROVED** [review]

**Comment by prince-chrismc** [comment]
My bot has this listed as "ready to merge" but master has been green for several hours... what's preventing it from merging?

**Comment by igormironchik** [comment]
At least one review is still needed, merging is blocked.

**Comment by igormironchik** [comment]
@lasote is set as pending reviewer...

**Comment by jgsogo** [comment]
It is the review by @uilianries the one blocking the merge. I will ping him.

**Review by uilianries — APPROVED** [review]
LGTM

**Comment by prince-chrismc** [comment]
> > My bot has this listed as "ready to merge" but master has been green for several hours... what's preventing it from merging?
>
> It is the review by @uilianries the one blocking the merge. I will ping him.

Think I found my discrepancy, Uilianries requested changes, approved, it was dismissed then approved

image

GitHub API shows "requested changes" and "dismissed" which I explictly handle in my bot [here](https://github.com/prince-chrismc/conan-center-index-pending-review/blob/f884a0528cb73786256a087408d03e793d35baef/pkg/pending_review/review.go#L87-L88)

Might be a feature request for @conan-center-bot but seems to be a rare corner case

**Comment by prince-chrismc** [comment]
> I cannot tell if it was approval or just a comment

:100: There's no way of knowing 🤷 

My hypothesis is the GitHub API changes an "approval" to a "dismissed" (nothing else) and changes a timestamp which is how it knows to add the event to the timeline (my screenshot).

 Sadly the GitHub docs suck https://docs.github.com/en/rest/reference/pulls#dismiss-a-review-for-a-pull-request so no confirmation 🤢 on the required preconditions... 

The only "dismissed" review is the [one from uilian](https://api.github.com/repos/conan-io/conan-center-index/pulls/5349/reviews/651588838). Since we can still the [requested changes and comments](https://api.github.com/repos/conan-io/conan-center-index/pulls/5349/reviews) I hope the premise for my bot is sound 🏛️ 

> If it was an approval it should have invalidated the previous blocking review

AFAIK is does **not** remove/edit "changes requested" reviews, this PR has [uilian's original one](https://api.github.com/repos/conan-io/conan-center-index/pulls/5349/reviews/651297910) still despite the [final approval](https://api.github.com/repos/conan-io/conan-center-index/pulls/5349/reviews/655546724)

This also holds for mathbunnyru reviews, both a requested change and an approval are listed

**Pull request merged**
